prusti-viper/src/encoder/snapshot_encoder.rs

```diff
@@ -177,33 +180,61 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
             | ty::TyKind::Uint(_)
             | ty::TyKind::Char
             | ty::TyKind::Bool => {
-                self.encode_snap_primitive(
+                self.encode_primitive(
                     self.encoder.encode_value_field(self.ty)
                 )?
             }
             ty::TyKind::Param(_) => {
-                self.encode_snap_generic()?
+                self.encode_generic()?
             }
-            ty::TyKind::Adt(adt_def, _) if !adt_def.is_box() => {
-                if adt_def.variants.len() != 1 {
-                    warn!("Generating equality tests for enums is not supported");
+            ty::TyKind::Tuple(_) => {
+                self.encode_tuple()?
+            }
+            ty::TyKind::Adt(adt_def, subst) if !adt_def.is_box() => {
+                if adt_def.is_struct() || adt_def.is_enum() {
+                    SnapshotAdtEncoder::new(
+                        &self,
+                        adt_def,
+                        subst,
+                    ).encode_snapshot()?
+                } else {
+                    unreachable!()
                 }
-                self.encode_snap_struct()?
             }
-            ty::TyKind::Tuple(_) => {
-                self.encode_snap_struct()?
+            ty::TyKind::Adt(adt_def, _) if adt_def.is_box() => { //TODO this breaks for lots of boxes. e.g. Box<Box<T>> or Box<i32>
+                let boxed_ty = self.ty.boxed_ty();
+                warn!("enocding boxed ty {:?}", boxed_ty);
+                match boxed_ty.kind() {
+                    ty::TyKind::Adt(adt_def, subst) => {
+                        if adt_def.is_struct() || adt_def.is_enum() {
+                            SnapshotAdtEncoder::new(
+                                &self,
+                                adt_def,
+                                subst,
+                            ).encode_snapshot()?
+                        } else {
+                            unreachable!()
+                        }
+                    }
+                    x => unimplemented!("{:?}", x),
+                }
             }
-
             x => unimplemented!("{:?}", x),
```

Comment:
It would be better to return `EncodingError::unsupported(..)` instead of calling `unimplemented!(..)`.